### ISSUE TITLE
Remove SliceConcatExt, add them as methods on [T].

### ISSUE THREAD

**Issue by bluss** [rust-lang/rfcs#1141]
Remove SliceConcatExt, add them as methods on [T].

We don't need extension traits for this anymore.

**Comment by gkoz** [comment]
Related PR: [Rename connect to join](https://github.com/rust-lang/rfcs/pull/1102).

BTW I'd very much like `Iterator` to have those methods instead. I know that would be less efficient than the current implementation. And maybe `impl Iterator for slice::Iter` could still salvage the more efficient implementation.

**Comment by bluss** [comment]
Moving a method from an unstable trait to an inherent impl is a minor breaking change I guess.

This quirky stuff compiles because concat is on a prelude trait:

```rust
<_>::concat(&["1", "2"][..])
```

**Comment by frewsxcv** [comment]
[Here's how Servo joins strings](https://github.com/servo/servo/blob/2049357d1453a43bb8ee31a4ba70d2610b93ff20/components/util/str.rs#L329-L337). It's nice that it works directly with iterators to avoid unnecessary allocations

**Comment by gkoz** [comment]
@frewsxcv 
Is `push_str` in that implementation not performing any allocations? Ones that `SliceConcatExt` [is avoiding](http://doc.rust-lang.org/src/collections/str.rs.html#88)?

**Comment by bluss** [comment]
Right, the slice version avoids reallocation while building the string by iterating through to sum up the exact capacity needed up front.

**Comment by frewsxcv** [comment]
I meant that if I have some other data structure like a `HashSet`, I'd have to iterate through all the string and collect them (causing an allocation to occur I think?) before calling `join` on it

**Comment by bluss** [comment]
Yes. Notice that servo could use SliceConcatExt's join *when it has a slice* for a much more efficient `join` operation. It's only unfortunate that we can't offer both the efficient slice version and the less efficient iterator version in a single method call. Specialization could help.
